### DISCUSSION TITLE
3D complex refractive index simulation using FD-BPM

### DISCUSSION THREAD

**Discussion by wesprit** [ankrh/BPM-Matlab#4] [Q&A] [answered]
Dear prof. Hansen
Hope my consulting finds you well!!
I'm interested in your work on the 2021 Optics Express artical "BPM-Matlab an open-source optical propagation simulation tool in MATLAB". This article introduces the FD-BPM applications of  fiber mode simulation in detail, which has benefited me a lot. I am very interested about this work, but I have something confused about this work so I was wondering if I could consult some questions?
1. In the open-source code download at https://gitlab.gbar.dtu.dk/biophotonics/BPM-Matlab, it is said that any complex fiber refractive index can be taken into the code for beam propagation simulation. However, when I try to bring a set of complex refractive index distribution values into the program 'Example 14' for calculation, it is found that the result is related to the parameters 'P.Nx_main' and 'P.Ny_main'. For example, if I set the parameters as 2120, "LP01 LP11 ..." modes can be found, see Fig. 1. 
![Fig  1](https://github.com/ankrh/BPM-Matlab/assets/49069211/5ccc41f8-d7da-44e5-a975-eb9d91fcce35)
However when I set the parameters as 1000, the modes found is completely different from 2120, see Fig. 2. 
![Fig  2](https://github.com/ankrh/BPM-Matlab/assets/49069211/5c4a42cc-c73f-46b2-9cf2-1cac72f9cba3)
Could you please explain the reason why the parameters 'P.Nx_main' and 'P.Ny_main' influence the mode finding results so much?
2. When I set the the parameters 'P.Nx_main' and 'P.Ny_main' as 2120, "LP01 LP11 ..." modes can be found, but the mode label only display as 'Mode 1, neff=1.48084',as shown in Fig. 1. I check the code in 'findMode' and found that it is related to parameters 'm' and 'l', but I still not understand why the label can not show LP01 while the LP01 mode picture can be found.
![Fig  1](https://github.com/ankrh/BPM-Matlab/assets/49069211/3cf6a4ac-7e25-4b98-8402-b8f28e41dd4d)
I attach the revelant section to this consulting relevant for your understanding. 
[Example14.txt](https://github.com/ankrh/BPM-Matlab/files/12517764/Example14.txt)
![complex RI](https://github.com/ankrh/BPM-Matlab/assets/49069211/0d2800ff-bee9-409d-bb70-013dced4486d)
Thank you very much!!
I fully appreciate that you have a busy schedule at the moment and therefore I cannot thank you enough for your time and effort. I am looking forward to your reply and I wish I can have a chat with you with respect. Many thanks in advance for your kind guidance.
                                                                                                                                                                                                   Best regards

**Comment by ankrh** [comment]
Hi Sijie Wang!

I also received an email from my co-developer Madhu about your issue. I will try to take a look at it in the coming days. It might be an issue of having too few pixels in the core region. The core seems to only occupy a small fraction of the simulation window size.

**Reply by wesprit** [reply]
Thank you for your reply!! 
The fiber used for mode simulation has a 25um core, 400um cladding and 530 um coating. Thus, the core seems to only occupy a small fraction in simulation window size. 
In example 14, the heat deposit was considered in refractive index calcluation, which i was writtern in calcRI function.
If you have any questions, please let me know.
Looking forward to hearing from you!!
Best regards

**Reply by ankrh** [reply]
Are you sure your RI is complex? It looks real to me...

**Reply by wesprit** [reply]
![fig3](https://github.com/ankrh/BPM-Matlab/assets/49069211/04503c32-0020-4b15-9e0d-00a4c1b4a185)
The distribution of RI in one △z slice is shown as Fig. 3 above. 
The left top gives the RI distribution along r axis while considering heat deposit. 
The others are the RI distribution of 1/4 fiber and entire fiber cross profile. 
The refractive index change caused by heat deposit is not very obvious. Fig. 3 illustrates this complex refractive index distribution in three different ways. Hoping it can answer your questions~

**Reply by ankrh** [reply]
I see, the refractive index is complicated but not complex in the mathematical sense - they have imaginary part zero (no absorption). I will try to look at it again soon but not sure exactly when I'll get time

**Reply by wesprit** [reply]
Dear prof. Hansen
Thank you for your response!
As the mode of optical fiber is mainly concentrated in the core part, I tried to calculate the refractive index distribution in R<50e-6m. Unfortunately, the findmode function still did not successfully identify the corresponding pattern after adding the number of pixels in the core.
I'm confused about your saying that "the refractive index is complicated but not complex in the mathematical sense - they have imaginary part zero (no absorption)." The refractive index in the fiber only considered the real part n. The BPM propagation method can not be used when setting the imaginary part as zero?
Thank you for your time!!

**Reply by ankrh** [reply, marked as answer]
When BPM-Matlab refers to complex refractive indices, it means what is described here: https://en.wikipedia.org/wiki/Refractive_index#Complex_refractive_index

But there's no problem with using a real refractive index distribution - in fact that's what most simulations use and it's simpler.

I tried to work my way through your model file. I found that BPM-Matlab did not correctly identify the refractive index as radially symmetric even though it was. I have made a new commit to the repository that should fix this issue and correctly label the modes as LP01, LP11e, LP11o etc.

Regarding correct calculation of the modes themselves, I recommend using a simulation window that is smaller and more close to the core. It only needs to be about twice the size of the core because your fields are still quite well confined to the core. I have made a modified and simplified version of your model file in which the modes are calculated much more accurately and much faster due to greatly reduced Nx and Ny.

Even if you want to launch these modes into a simulation with a larger window size, you can still calculate the modes with the smaller window initially. BPM-Matlab will automatically resample the found modes to any simulation window size and resolution you may subsequently specify.

[WongSijeiModelFile.zip](https://github.com/ankrh/BPM-Matlab/files/12643185/WongSijeiModelFile.zip)

By the way, your script runs very slowly, probably because your style of programming is more like C-style programming and uses a lot of loops to iterate over your arrays one element at a time. You'll be able to make your scripts run much faster and look more elegant by using more MATLAB-style code, like what I have introduced in the attached model file :-)

**Comment by wesprit** [comment]
Dear prof. Hansen
Thank you for taking time out of your busy schedule to help me~ I really appreciate it.
I have carefully read the file you give to me. It seems the problem why my code can't recognize the mode number is on RI modelling. 
The code I write is time sonsuming to run and causing so many problem in simulation. I will continue to study the BPM-MATLAB code you write and improve my programming skills.
Best regards!

**Reply by ankrh** [reply]
No worries. Your programming skills are fine, but every language is a bit different and it takes a little time to learn how to make use of MATLAB's features and notation.

Often, operating on arrays would be done with loops in C but without loops in MATLAB.
